File: main/sort-array-by-parity-ii/sort-array-by-parity-ii.py
class Solution:
    def sortArrayByParityII(self, a):
        """
        :type A: List[int]
        :rtype: List[int]
        """
        evens = (x for x in a if x % 2 == 0)
        odds = (x for x in a if x % 2 == 1)
        return list(itertools.chain.from_iterable(zip(evens, odds)))

File: main/sort-array-by-parity-ii/sort-array-by-parity-ii.cpp
class Solution {
public:
    static vector<int> sortArrayByParityII(vector<int>& a) noexcept;
};

vector<int> Solution::sortArrayByParityII(vector<int>& a) noexcept
{
    const auto n = static_cast<int>(a.size());
    vector<int> b (n);
    auto i = 0, j = 0;

    for (auto k = 0; k < n; k += 2) {
        while (a[i] % 2 != 0) ++i;
        b[k] = a[i++];

        while (a[j] % 2 == 0) ++j;
        b[k + 1] = a[j++];
    }

    return b;
}
